Storm-POMDP update to latest version

AlexBork opened the pull request:
This PR contains the updates to Storm-POMDP with regards to the quantitative analysis. 
The main functionality added:
- Improvements to cut-offs for the under-approximation
- Inclusion of the clipping approach as described in the TACAS 2022 paper
- Extension of the implementation to interact with PAYNT for the SAYNT approach (to appear at CAV 2023)
- Streamlining of the AR-loop implementation: the one-shot computation is now fully integrated

Furthermore the PR brings large parts of Storm-POMDP to the latest standards used in the rest of Storm.

Thanks to @SvStein for helping out with a lot of the refactoring!

AlexBork commented:
My suggestion is to add Storm-POMDP to the formatting check and run the auto-formatting after the merge. That should make the review easier.

volkm commented on README.md:
```diff
@@ -1,10 +1,9 @@
-[![GitHub release](https://img.shields.io/github/release/moves-rwth/storm.svg)](https://github.com/moves-rwth/storm/releases/)
-[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1181896.svg)](https://doi.org/10.5281/zenodo.1181896)
-
+<!-- [![Build Status](https://github.com/moves-rwth/storm/workflows/Build%20Test/badge.svg)](https://github.com/moves-rwth/storm/actions) -->
```
This should be reverted.

AlexBork replied:
Done

volkm commented on CMakeLists.txt:
```diff
@@ -581,4 +581,4 @@ install(FILES ${CMAKE_BINARY_DIR}/stormConfig.install.cmake DESTINATION ${CMAKE_
 
-include(StormCPackConfig.cmake)
\ No newline at end of file
+include(StormCPackConfig.cmake)
```
Can we revert this?

AlexBork replied:
Done.

volkm commented on src/storm/modelchecker/prctl/helper/SparseMdpPrctlHelper.cpp:
```diff
@@ -379,7 +379,9 @@ SparseMdpHintType<ValueType> computeHints(Environment const& env, SolutionType c
+        if (hint.isEmpty()) {
+            STORM_LOG_TRACE("Warn A non-empty hint was provided, but its information will be disregarded.");
+        }
     }
```
What is the reason for this change? Too much output otherwise?

AlexBork replied:
Exactly, in particular because it is not really something a user should be concerned about.

AlexBork commented:
Yes, it does. I originally applied the auto-format, but got confused why it didn't apply to the POMDP code so I reverted. I can apply formatting now or we can merge and then also apply formatting to the POMDP code, depending on whether it is acceptable to have the formatting check fail in the meantime.

tquatmann commented on src/storm-pomdp/modelchecker/PreprocessingPomdpValueBoundsModelChecker.cpp:
```diff
@@ -0,0 +1,321 @@
+        namespace modelchecker {
+            template <typename ValueType>
+            PreprocessingPomdpValueBoundsModelChecker<ValueType>::PreprocessingPomdpValueBoundsModelChecker(storm::models::sparse::Pomdp<ValueType> const& pomdp, storm::solver::MinMaxMethod minMaxMethod) : pomdp(pomdp) {
+                // Setup checker environment
+                mcEnvironment = storm::Environment();
+                mcEnvironment.solver().setLinearEquationSolverType(storm::solver::EquationSolverType::Topological, true);
+                mcEnvironment.solver().minMax().setMethod(minMaxMethod);
+                mcEnvironment.solver().minMax().setPrecision(storm::utility::convertNumber<storm::RationalNumber>(1e-9));
+                mcEnvironment.solver().topological().setUnderlyingEquationSolverType(storm::solver::EquationSolverType::Eigen);
```
Why do we require topological + Eigen for DTMCs and (non-topological) SVI for MDPs?
Maybe, what we actually want here is something like `setForceExact` or `setForceSoundness` in `SolverEnvironment` ?
In any case, it would be nice to elaborate a bit on this.

Usually, we pass a `storm::Environment cons& env` to all the methods that need an environment, i.e., it should be added as the first argument of `getValueBounds` and `getExtremeValueBound`.  Then, the calling method needs to take care of configuring the environment which makes things a bit more transparent.

There might be reasons to do it differently here that I don't see right now, though.

tquatmann commented on src/storm-pomdp/modelchecker/BeliefExplorationPomdpModelCheckerOptions.h:
```diff
@@ -36,9 +46,18 @@ namespace storm {
+
+                storm::builder::ExplorationHeuristic explorationHeuristic = storm::builder::ExplorationHeuristic::BreadthFirst;
+
+                storm::solver::MinMaxMethod preProcMinMaxMethod = storm::solver::MinMaxMethod::SoundValueIteration;
```
I suggest to have a `storm::Environment preProcEnvironment` instead if this really needs to be configured "from the outside"

AlexBork commented on src/storm-pars/transformer/SparseParametricDtmcSimplifier.cpp:
```diff
@@ -34,8 +34,9 @@ namespace storm {
-            storm::storage::BitVector maybeStates = reachableGreater0States & ~statesWithProbability01.second;
-            
+            //storm::storage::BitVector maybeStates = reachableGreater0States & ~statesWithProbability01.second;
+            storm::storage::BitVector maybeStates = ~(statesWithProbability01.second | statesWithProbability01.first);
```
Yes, I think that was the point. However, the native parametric transformation checking is not supposed to be part of the release, so I'll revert this change for now. I'll also keep your comment in mind for later implementations.

tquatmann reviewed (commented):
Besides my two minor comments, I approve 👍 

[I forward declared the environment to optimise compile times a little.]

AlexBork commented:
I incorporated Tim's comments and fixed a small bug triggered by the new VI implementation in conjunction with infinite reward values. The current state of the main branch is also merged.
